conformance/lang/expressions/let_expr.balt

```diff
@@ -0,0 +1,612 @@
+    int i = 2;
+    io:println(let int i = 2 in i); // @error redeclared symbol 'i'
+}
+
+Test-Case: error
+Description: Test let expression when the type of let variables does not compatible with let expression.
+Labels: function-call-expr, int, let-expr, optional-type, string, string-literal, var
+
+function errorFunction() {
+    io:println(let string i = "2", int j = 3 in i + j); // @error operator '+' not defined for 'string' and 'int'
+    io:println(let int i = 2 in foo(i)); // @error incompatible types: expected 'string?', found 'int'
+    io:println(let var i = 2 in foo(i)); // @error incompatible types: expected 'string?', found 'int'
+}
+
+function foo(string? s) returns string? {
+    return s;
+}
+
+Test-Case: error
+Description: Test let expression when when the let variables are incorrect.
+Labels: int, let-expr, string-literal
+
+function errorFunction() {
+    io:println(let int i = "not int" in i); // @error incompatible types: expected 'int', found 'string'
+}
+
+Test-Case: output
+Description: Test let expression being an isolated expression, when the expression following `in` is isolated.
+             A reference in that expression to a variable bound by a let-var-decl will be treated as isolated
+             if the expression initializing the variable in the let-var-decl is isolated.
+Labels: function-call-expr, int, let-expr
+
+isolated int x = 2;
+isolated int y = 4;
+
+isolated function bar(int i) returns int {
+    return i;
+}
+
+isolated function init() {
+    lock {
+        x = let int i = 20 in i;
+        io:println(x); // @output 20
+        x = let int i = 22 in bar(i);
+        io:println(x); // @output 22
+    }
+
+    lock {
+        int x1 = let int i = y in bar(i);
+        io:println(x1); // @output 4
+        int x2 = let int i = bar(200) in bar(i);
+        io:println(x2); // @output 200
+    }
+}
+
+Test-Case: error
+Description: Test error cases when let expression is assigned to an isolated variable.
+             Non isolated expressions cannot be used to initialize let-var-decl when expression following in expected to be isolated.
+Labels: array-type, function-call-expr, let-expr
+
+isolated int[] x = [];
+
+function foo() {
+    lock {
+        int[] y = [23];
+        x = let int[] i = y in i; // @error the let expression is not isolated because the expression that initializes variable i is not isolated
+    }
+}
+
+function errorFunction() {
+    foo();
+}
+
+Test-Case: output
+Description: Test behaviour if the typed-binding-pattern uses var.
+             The type of the variable should be inferred from the precise static type of the expression following =.
+Fail-Issue: ballerina-platform/ballerina-lang#35527
+Labels: function-call-expr, let-expr, module-type-defn, singleton-type, string-literal, var
+
+type X "char_X";
+
+function foo(X x) returns boolean {
+    return x == "char_X";
+}
+
+function init() {
+    _ = let var x = "char_X" in foo(x); // @output true
+}
```

Comment:
Do we have a test to cover 

> Since expressions cannot modify variables, the variables bound in a let-var-decl are implicitly final.

Came across https://github.com/ballerina-platform/ballerina-lang/issues/37595